components/service/glean/src/main/java/mozilla/components/service/glean/storages/StringListsStorageEngine.kt

```diff
@@ -0,0 +1,83 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.storages
+
+import android.annotation.SuppressLint
+import mozilla.components.service.glean.CommonMetricData
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * This singleton handles the in-memory storage logic for string lists. It is meant to be used by
+ * the Specific String List API and the ping assembling objects. No validation on the stored data
+ * is performed at this point: validation must be performed by the Specific String List API.
+ *
+ * This class contains a reference to the Android application Context. While the IDE warns
+ * us that this could leak, the application context lives as long as the application and this
+ * object. For this reason, we should be safe to suppress the IDE warning.
+ */
+@SuppressLint("StaticFieldLeak")
+internal object StringListsStorageEngine : StringListsStorageEngineImplementation()
+
+internal open class StringListsStorageEngineImplementation(
+    override val logger: Logger = Logger("glean/StringsStorageEngine")
+) : GenericScalarStorageEngine<List<String>?>() {
+
+    companion object {
+        // Maximum length of any list
+        const val MAX_LIST_LENGTH_VALUE = 20
+    }
+
+    @SuppressWarnings("unchecked")
+    override fun deserializeSingleMetric(value: Any?): List<String>? {
+        return value as? List<String>
+    }
+
+    /**
+     * Appends to an existing string list in the desired stores.  If the store or list doesn't exist
+     * then it is created and added to the desired stores.
+     *
+     * @param metricData object with metric settings
+     * @param value the string list value to add
+     */
+    @Synchronized
+    fun add(
+        metricData: CommonMetricData,
+        value: String
+    ) {
+        // Use a custom combiner to add the string to the existing list rather than overwriting
+        super.recordScalar(metricData, listOf(value)) { currentValue, newValue ->
+            currentValue?.let {
+                if (it.count() + value.count() > MAX_LIST_LENGTH_VALUE) {
+                    logger.warn("${metricData.category}.${metricData.name} - " +
+                                "discarding extra values. List limit $MAX_LIST_LENGTH_VALUE " +
+                                "reached.")
+                }
+
+                it + newValue!!.take(MAX_LIST_LENGTH_VALUE - it.count())
```

Comment:
The scary `!!` should not be needed anymore if we remove the nullability of the type in the class definition, as pointed out above.

Reply:
Not sure why there wasn't a comment box above for the `@SuppressWarnings()` but I get the following error when I try changing the code to the following:
```
    override fun deserializeSingleMetric(value: Any?): List<String>? =
        if (value is List<String>) value as List<String> else null
```
`Cannot check for instance of erased type: List<String>`
The IDE suggests changing it to `if (value is List<*>)` but I'm afraid that may allow for deserialization of types other than Lists of Strings.

It also complains that the cast is not necessary, I'm guessing because the `if` allows for type inference.

Reply:
I just tried it like this:
```
    override fun deserializeSingleMetric(value: Any?): List<String>? =
        if (value is List<*>) value as List<String> else null
```
and I get the `Unchecked cast: Any? to List<String>` warning again